swe_roe/ex2b.c

```diff
@@ -1433,6 +1469,77 @@ PetscErrorCode AddBuildings(RDyApp app) {
   PetscFunctionReturn(0);
 }
 
+/// @brief Reads a PETSc Vec in binary format specified via -boundary_edge_type_file and marks
+///        boundary edge type
+/// @param [in] app An application context
+/// @return 0 on success, or a non-zero error code on failure
+PetscErrorCode MarkBoundaryEdgeType(RDyApp app) {
+  PetscFunctionBeginUser;
+
+  RDyMesh  *mesh  = &app->mesh;
+  RDyEdges *edges = &mesh->edges;
+
+  PetscViewer viewer;
+  PetscCall(PetscViewerBinaryOpen(app->comm, app->boundary_edge_type_file, FILE_MODE_READ, &viewer));
+
+  Vec natural, global, local;
+
+  // Create Vecs
+  PetscCall(DMPlexCreateNaturalVector(app->auxdm, &natural));
+  PetscCall(DMCreateGlobalVector(app->auxdm, &global));
+  PetscCall(DMCreateLocalVector(app->auxdm, &local));
+
+  // Load the data
+  PetscCall(VecLoad(natural, viewer));
+
+  // Scatter from natural-to-global order
+  PetscCall(DMPlexNaturalToGlobalBegin(app->auxdm, natural, global));
+  PetscCall(DMPlexNaturalToGlobalEnd(app->auxdm, natural, global));
+
+  // Scatter from global-to-local order
+  PetscCall(DMGlobalToLocalBegin(app->auxdm, global, INSERT_VALUES, local));
+  PetscCall(DMGlobalToLocalEnd(app->auxdm, global, INSERT_VALUES, local));
+
+  PetscScalar *local_ptr;
+  PetscCall(VecGetArray(local, &local_ptr));
+
+  for (PetscInt icell = 0; icell < mesh->num_cells; icell++) {
+    // Is a boundary condition being applied on the cell?
+    if ((PetscInt)local_ptr[icell] == PRESCRIBED_HEAD || (PetscInt)local_ptr[icell] == CRITICAL_OUTFLOW) {
+      PetscBool edge_found = PETSC_FALSE;
+
+      // Loop over all boundary edges to identify the corresponding edge
+      for (PetscInt ii = 0; ii < mesh->num_boundary_edges; ii++) {
+        PetscInt iedge      = edges->boundary_edge_ids[ii];
+        PetscInt cellOffset = edges->cell_offsets[iedge];
+        PetscInt l          = edges->cell_ids[cellOffset];
+
+        // Is this edge corresponding the to the icell-th cell?
+        if (l == icell) {
+          edge_found                     = PETSC_TRUE;
+          edges->boundary_edge_types[ii] = (PetscInt)local_ptr[icell];
+          break;
+        }
+      }
```

Comment:
Probably a faster method is to find the BC edges connected to ghost cells, then you don't need to do this for loop. But it should be OK since the number of boundary edges will not be very significant.

Reply:
The BC is applied on the edge of an internal cell and not a ghost cell.